cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -106,7 +106,12 @@ struct ValueConverter<Type, enable_if_integer<Type>> {
 
   static inline Result<ValueType> FromPython(PyObject* obj) {
     ValueType value;
-    RETURN_NOT_OK(internal::CIntFromPython(obj, &value));
+    arrow::Status s_ = internal::CIntFromPython(obj, &value);
+    if (!s_.ok() && !internal::PyIntScalar_Check(obj)) {
+      return internal::InvalidValue(obj, "tried to convert to int");
+    } else {
+      RETURN_NOT_OK(s_);
```

Comment:
What are the cases that this couldn't be converted, but that `obj` *is* an integer? When the integer is too big to fit in a C int?

Reply:
Yes, and also when converting a negative integer to a `uint`:
```
pa.scalar(-1, type='uint8')
```
No other tests are touched if I recompile without this check